File: leetcode/problems_list/125-ValidPalindrome/solution.js
/**
 * @param {string} s
 * @return {boolean}
 */
// Time Complexity: O(n)
// Space Complexity: O(1)
var isPalindrome = function (s) {
  let cleanString = s.toLowerCase().replace(/[^A-Za-z0-9]/g, "");
  let rIndex = cleanString.length - 1;
  for (const curr of cleanString) {
    if (curr !== cleanString[rIndex]) return false;
    rIndex--;
  }
  return true;
};

/**
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *   let cleanString = s.toLowerCase().replace(/[^A-Za-z0-9]/g, "");
 *   let reverseString = cleanString.split("").reverse().join("");
 *   return cleanString === reverseString;
 */

File: leetcode/problems_list/125-ValidPalindrome/solution.py

# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
    def isPalindrome(self, s: str) -> bool:
        clean_string = "".join(filter(str.isalnum, s.lower()))
        right_index, left_index = len(clean_string) - 1, 0
        while left_index < right_index:
            if clean_string[left_index] != clean_string[right_index]:
                return False
            left_index += 1
            right_index -= 1
        return True


'''
# Time complexity: O(n)
# Space complexity: O(n)
    def isPalindrome(self, s: str) -> bool:
        clean_string = "".join(filter(str.isalnum, s.lower()))
        print(clean_string, clean_string[::-1])
        return clean_string == clean_string[::-1]
'''
